neqo-http3/tests/webtransport/mod.rs

```diff
@@ -0,0 +1,205 @@
+                    },
+                ) => {
+                    assert!(
+                        headers
+                            .iter()
+                            .any(|h| h.name() == ":method" && h.value() == "CONNECT")
+                            && headers
+                                .iter()
+                                .any(|h| h.name() == ":protocol" && h.value() == "webtransport")
+                    );
+                    session.response(accept).unwrap();
+                    wt_server_session = Some(session);
+                }
+                Http3ServerEvent::Data { .. } => {
+                    panic!("There should not be ane data events!");
+                }
+                _ => {}
+            }
+        }
+
+        self.exchange_packets();
+        (wt_session_id, wt_server_session)
+    }
+
+    fn create_wt_session(&mut self) -> (u64, WebTransportRequest) {
+        let (wt_session_id, wt_server_session) = self.negotiate_wt_session(true);
+        let wt_session_negotiated_event = |e| {
+            matches!(
+                e,
+                Http3ClientEvent::WebTransport(WebTransportEvent::WebTransportSession(stream_id)) if stream_id == wt_session_id
+            )
+        };
+        assert!(self.client.events().any(wt_session_negotiated_event));
+
+        (wt_session_id, wt_server_session.unwrap())
+    }
+
+    fn exchange_packets(&mut self) {
+        let mut out = None;
+        loop {
+            out = self.client.process(out, now()).dgram();
+            out = self.server.process(out, now()).dgram();
+            if out.is_none() {
+                break;
+            }
+        }
+    }
+
+    pub fn cancel_session_client(&mut self, wt_stream_id: u64) {
+        self.client
+            .cancel_fetch(wt_stream_id, Error::HttpNoError.code())
+            .unwrap();
+        self.exchange_packets();
+    }
+
+    pub fn check_session_closed_event_client(
+        &mut self,
+        wt_session_id: u64,
+        expected_error: Option<AppError>,
+    ) {
+        let wt_session_closed = |e| {
+            matches!(
+                e,
+                Http3ClientEvent::WebTransport(WebTransportEvent::WebTransportSessionClosed{stream_id, error}) if stream_id == wt_session_id && error == expected_error
+            )
+        };
+        assert!(self.client.events().any(wt_session_closed));
+    }
+
+    pub fn cancel_session_server(&mut self, wt_session: &mut WebTransportRequest) {
+        wt_session.cancel_fetch(Error::HttpNoError.code()).unwrap();
+        self.exchange_packets();
+    }
+
+    pub fn check_session_closed_event_server(
+        &mut self,
+        wt_session: &mut WebTransportRequest,
+        expected_error: Option<AppError>,
+    ) {
+        let wt_session_closed = |e| matches!(e, Http3ServerEvent::WebTransport(WebTransportServerEvent::WebTransportSessionClosed{session, error}) if session.stream_id() == wt_session.stream_id() && error == expected_error);
```

Comment:
This line is a bit too long, I think (same for the above).  Maybe you can make an associated/static function for this and try to tame it some:

```rust
fn session_closed_client(e: &Http3ClientEvent, id: u64, expected_error: &Option<AppError>) -> bool {
    if let Http3ClientEvent::WebTransport(e) = e {
        if let WebTransportEvent::WebTransportSessionClosed { stream_id, error } = e {
            *stream_id == id && error == expected_error
        } else {
            false
        }
    } else {
        false
    }
}
```
Longer, but maybe easier to read.  Don't know what clippy would think of that though.